### ISSUE TITLE
Angular Interceptor unable to calculate correct time of API response, when there is multiple request at the same time

### ISSUE THREAD

**Issue by mdarif-k** [angular/angular#38037]
Steps to reproduce the issue:

The First API call response is coming in 600ms.
within the first API call response time, another 4 API request is going to the interceptor and this process takes another 1 second.
The issue is the interceptor is handling the request as a priority but the First API call response is ready in 600ms but it is returning the response once All the request processes, which means around 1.6 seconds which is not the correct calculated time.

The logic to calculate the time of API response.

<pre><code>

const startTime = new Date().getTime();

next.handle(req).pipe(
  map((event) => {
     const endTime = new Date().getTime();
     const diff = (endTime - startTime)/1000 + 'Seconds'; 
     return event;
  })
)

</code></pre>

**Comment by JoostK** [comment]
I am sorry, but I don't understand what you mean exactly and it's not clear to me how this would be a bug in Angular? Could you try to clarify your problem?

**Comment by mdarif-k** [comment]
Hi JoostK, 
Thanks for your comment.
Let me try to explain the issue one more time using the below image from the network Tab.

![image](https://user-images.githubusercontent.com/22832843/87407970-7dac2000-c5e0-11ea-8952-6eadaa4e30cd.png)


If you will see the attached network tab image, where the giftCardCompativility call is response is coming in 658 ms and before 658 ms there are another 3 fetchRules requests that are going to an interceptor.

The issue is the interceptor giving a response of giftcardCompatavility API after taking care of all incoming requests which increasing the calculation time of response.

If there are no multiple requests than the interceptor is calculating almost the same time as the network tab.

**Comment by JoostK** [comment]
As I understand it, you're measuring durations in an interceptor and you're seeing this duration span multiple requests, whereas you expect only the single request to be measured. This would occur if additional requests are made from other interceptors, which I can't tell based on the code you provided.

Unfortunately it's still not clear to me how this could be a bug in Angular, so I'd ask you to provide a reproduction of the problem or _at least_ the code that makes these requests (`giftcardCompatibility` and `fetchRules`) together with the relevant interceptors to better understand problem you're having.

**Comment by mdarif-k** [comment]
I can not share the exact code since it is in the client machine.

sure. I will create a dummy project and try to reproduce there and share the repository.

Thanks for your help JoostK

**Comment by sacgrover** [comment]
@mdarif-k 

Here is my understanding:
First, it is not related to angular. this is the general behavior of java script.

In your case, subscribe call is asyn for giftCardCompativility but other calls are in sync. So all the sync code execution is placed in a same main flow whereas subscribe/setTimeout/async functions are pushed to the event loop. That's why when you have other api calls they called first in main flow execution before calling the subscribe of giftCardCompativility. All the pipe and map you added depends on subscribe call. they generate results just before subscribe call.

I hope, this will make sense.. :)

**Comment by mdarif-k** [comment]
Hi Sacgrover,

Thanks for your comment and it helps to understand the issue. But I am handling all the calls with a subscription.

giftCardCompativility call is inside the other subscription but all the other calls having their own subscription means they are not inside any subscription.

But giftCardCompativility call is going first and other calls are later. In this case also giftCardCompativility will execute later?

**Comment by sacgrover** [comment]
@mdarif-k  Yes, Async calls will never give you a guarantee that which code will execute first and you can not compare browser call response timing with your code execution timing in this way. They will be most of the time different.

If it answered your question then please close this issue as this is not angular related. Instead post on stack overflow for possible solutions.

**Comment by mdarif-k** [comment]
Thank you @sacgrover and @JoostK for help, I am closing this issue. It is very difficult to replicate in another project because I need backend support.

**Comment by mdarif-k** [comment]
I got the issue, on continue from one page to another page I have submitted task API. this submit task API has another API call giftcardcompatability, we are calling this API in async/subscribe mode and we are navigating to the next page based on submitTask API call. When the next page start loads that time lot of Splunk call going to the interceptor and all Splunk call in sync/topromise so it is getting registered in the main loop since giftcardcompatability call is async call so that it is getting registered in event loop and event loop executes after the main loop so the response is coming after sync calls.
